vortex-serde/src/layouts/read/stream.rs

```diff
@@ -74,107 +74,127 @@ impl<R: VortexReadAt> LayoutBatchStream<R> {
 
 type StreamStateFuture<R> = BoxFuture<'static, VortexResult<(R, Vec<(MessageId, Bytes)>)>>;
 
-#[derive(Default)]
+/// State of vortex file stream
+///
+/// The flow starts with `AddSplits` which produces all horizontal row splits in the file
+/// Main read loop goes from `NextSplit` -> `Filter` (if there's filter) -> `Read`
+/// `Filter` and `Read` states transition to `Reading` when they're blocked on an io operation which resumes back to
+/// the previous state.
 enum StreamingState<R> {
-    #[default]
-    Init,
-    FilterInit,
-    Reading(StreamStateFuture<R>),
-    FilterReading(StreamStateFuture<R>),
-    Decoding(Array),
+    AddSplits,
+    NextSplit,
+    Filter,
+    Read,
+    Reading(StreamStateFuture<R>, bool),
     Error,
 }
 
-impl<R: VortexReadAt + Unpin + Send + 'static> Stream for LayoutBatchStream<R> {
+impl<R: VortexReadAt + Unpin + 'static> Stream for LayoutBatchStream<R> {
     type Item = VortexResult<Array>;
 
     fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
         loop {
             match &mut self.state {
-                StreamingState::Init => {
-                    if let Some(read) = self.layout_reader.read_next()? {
-                        match read {
-                            ReadResult::ReadMore(messages) => {
-                                let reader = self.input.take().ok_or_else(|| {
-                                    vortex_err!("Invalid state transition - reader dropped")
-                                })?;
-                                let read_future = read_ranges(reader, messages).boxed();
-                                self.state = StreamingState::Reading(read_future);
-                            }
-                            ReadResult::Batch(a) => self.state = StreamingState::Decoding(a),
-                        }
-                    } else {
+                StreamingState::AddSplits => {
+                    let mut splits = BTreeSet::new();
+                    splits.insert(self.row_count as usize);
+                    self.filter_reader
+                        .as_mut()
+                        .map(|fr| fr.add_splits(0, &mut splits))
+                        .unwrap_or_else(|| {
```

Comment:
Why is this unwrap_or_else? Shouldn't we just add splits from both of the readers always?